### PR TITLE
Yuzhe/mapmatching refactor

### PR DESCRIPTION
**Pull request by yuzheyan**
# Issue
Refactoring some header files in meili:
changes including 
1. removing redundant blank lines
2. adding default values to data members to prevent potential bugs.
3. moving function implementation into .cc to improve readability.
4. simplify some branching logic
5. reordering and group the member function/member variables.  
    function grouping: constructor, destructor, logic function, helpers(getter, setters),
    variable grouping:  pointer/&, std containers, user defined types, build in primary types. 
6. redeclare some private const function to private static function.
7. remove inline keyword/replace inline keyword with constexpr
8. remove useless virtual keywords to reduce virtual table size.
9. reordering the header file including sequence 
    1st defination head, 2nd standard library header, 3rd boost head, 4th within project headers

## Requirements / Relations

 Link any requirements here. Other pull requests this PR is based on?

### CODE REVIEW

File: src/meili/viterbi_search.cc [hunk 1]
```diff
@@ -192,23 +192,28 @@ template <bool Maximize> bool NaiveViterbiSearch<Maximize>::AddStateId(const Sta
...
-  const auto removed = IViterbiSearch::RemoveStateId(stateid);
+  const bool removed = IViterbiSearch::RemoveStateId(stateid);
   if (!removed) {
     return false;
   }
   // remove it from columns
   auto& column = states_[stateid.time()];
   const auto it = std::find(column.begin(), column.end(), stateid);
+  if (it == column.end()) {
+    throw std::runtime_error{"could not find StateId in column"};
```
[the remark this answers is not in the document]
Comment (yuzheyan) [reply src/meili/viterbi_search.cc]: Our code previously did not check whether it exist or not. vector.erase(vecor.end()) is undefined behavior. If we didn't have crashes from here, maybe to set return as false is safe.

File: valhalla/meili/viterbi_search.h [hunk 2]
```diff
@@ -74,294 +51,133 @@ class StateIdIterator : public std::iterator<std::forward_iterator_tag, StateId>
...
   IViterbiSearch& vs_;
-
-  StateId::Time time_;
-
-  StateId stateid_;
-
-  bool allow_breaks_;
-
-  void Next();
+  StateId::Time time_{kInvalidTime};
```
[the remark this answers is not in the document]
Comment (yuzheyan) [reply valhalla/meili/viterbi_search.h]: Old code has two user-defined constructors. The first one takes in time as parameter and initialize it in constructor. The second one left it with uninitialized.  I would assume it is a safe behavior to set it to kInvalid when the second constructor is called.

File: src/meili/viterbi_search.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (yuzheyan) [reply src/meili/viterbi_search.cc]: I found out our code previously throws an logic error saying stateId must exist in the column. I added back to make it consistent with the previous design.

[the remark this answers is not in the document]
Comment (kevinkreiser) [reply src/meili/viterbi_search.cc]: ok cool thanks!

**Pull request merged**